ledger/complete/mtrie/node/node.go

```diff
@@ -122,12 +122,98 @@ func NewInterimNode(height int, lchild, rchild *Node) *Node {
 	return n
 }
 
+// deepCopy
+func (n *Node) deepCopy() *Node {
+	var h hash.Hash
+	copy(h[:], n.hashValue[:])
+
+	var p ledger.Path
+	copy(p[:], n.path[:])
+
+	var lChildCopy, rChildCopy *Node
+	if n.lChild != nil {
+		lChildCopy = n.lChild.deepCopy()
+	}
+	if n.rChild != nil {
+		rChildCopy = n.rChild.deepCopy()
+	}
+
+	return &Node{
+		lChild:    lChildCopy,
+		rChild:    rChildCopy,
+		height:    n.height,
+		path:      p,
+		payload:   n.payload.DeepCopy(),
+		maxDepth:  n.maxDepth,
+		regCount:  n.regCount,
+		hashValue: h,
+	}
+}
+
 // computeAndStoreHash computes the node's hash value and
 // stores the result in the nodes internal `hashValue` field
 func (n *Node) computeAndStoreHash() {
 	n.hashValue = n.computeHash()
 }
 
+// Compactify an expanded leaf into its most concise representation. The
+// compactified representation of a default node is `nil`. For a node that only has a
+// _single_ child that is itself a leaf, this method returns a new, fully compactified leaf.
+// Returns:
+//  * (n, false) if the node cannot be compactified, we return the original node `n`
+//  * (cn, true) if the node can be compactified, where cn is a newly created compactified leaf
+func (n *Node) Compactify() (*Node, bool) {
+	if n.IsLeaf() {
+		return n, false
+	}
+
+	// if leaf return it as is
+	// if non leaf bubble up
+	lChildEmpty := true
+	rChildEmpty := true
+	if n.lChild != nil {
+		lChildEmpty = n.lChild.IsDefaultNode()
+	}
+	if n.rChild != nil {
+		rChildEmpty = n.rChild.IsDefaultNode()
+	}
+	if rChildEmpty && lChildEmpty {
+		// is like a leaf
+		return NewLeaf(ledger.Path{}, ledger.EmptyPayload(), n.height), true
+	}
+
+	// if childNode is non empty
+	if !lChildEmpty && rChildEmpty && n.lChild.IsLeaf() {
+		// TODO merge these two methods
+		l := n.lChild.deepCopy()
+		l.hashValue = hash.HashInterNode(l.hashValue, ledger.GetDefaultHashForHeight(l.height))
+		l.height = l.height + 1
+
+		// bubble up all children and return as parent
+		return l, true
+	}
+	if lChildEmpty && !rChildEmpty && n.rChild.IsLeaf() {
+		// TODO merge these two methods
+		r := n.rChild.deepCopy()
+		r.hashValue = hash.HashInterNode(ledger.GetDefaultHashForHeight(r.height), r.hashValue)
+		r.height = r.height + 1
+		// bubble up all children and return as parent
+		return r, true
+	}
+
+	// else no change needed
+	return n, false
+}
+
+func (n *Node) IsDefaultNode() bool {
+	//  TODO make this optimize by caching if the node is a default node
```

Comment:
From the calls to `IsDefaultNode` it doesn't seem to me that the same node is called more than once (but I might me wrong). So I guess caching is not required (also the test itself is cheap)
```suggestion
```